src/resolve_links.cpp

```diff
@@ -130,6 +143,41 @@ void resolve_links::match(std::shared_ptr<const torrent_info> const& ti
 	}
 
 }
+void resolve_links::match_v2(std::shared_ptr<const torrent_info> const& ti
+	, std::string const& save_path)
+{
+	file_storage const& fs = ti->files();
+	m_file_sizes.reserve(aux::numeric_cast<std::size_t>(fs.num_files()));
+	for (auto const i : fs.file_range())
+	{
+		// for every file in the other torrent, see if we have one that match
+		// it in m_torrent_file
+		if (fs.pad_file_at(i)) continue;
+
+		std::int64_t const file_size = fs.file_size(i);
+
+		auto const range = m_file_sizes.equal_range(file_size);
+		for (auto iter = range.first; iter != range.second; ++iter)
+		{
+			TORRENT_ASSERT(iter->second >= file_index_t(0));
+			TORRENT_ASSERT(iter->second < m_torrent_file->files().end_file());
+
+			// if we already have found a duplicate for this file, no need
+			// to keep looking
+			if (m_links[iter->second].ti) continue;
+
+			if (fs.root(i) == m_torrent_file->files().root(iter->second)) {
+				m_links[iter->second].ti = ti;
+				m_links[iter->second].save_path = save_path;
+				m_links[iter->second].file_idx = i;
+				// since we have a duplicate for this file, we may as well remove
+				// it from the file-size map, so we won't find it again.
+				m_file_sizes.erase(iter);
+				break;
+			}
+		}
+	}
+}
```

Comment:
This function can be a lot simpler. All you need to compare are the file root hashes.

I would suggest building a map of file roots to indices up front though, to ensure that even torrents with a lot of files will be able to run this match without bogging down the system. e.g. `std::unordered_map<sha256_hash, file_index_t>`.